arangod/Cluster/RebootTracker.cpp

```diff
@@ -0,0 +1,330 @@
+  }
+}
+
+CallbackGuard RebootTracker::callMeOnChange(RebootTracker::PeerState const& peerState,
+                                            RebootTracker::Callback callback,
+                                            std::string callbackDescription) {
+  MUTEX_LOCKER(guard, _mutex);
+
+  auto const rebootIdIt = _rebootIds.find(peerState.serverId());
+
+  // We MUST NOT insert something in _callbacks[serverId] unless _rebootIds[serverId] exists!
+  if (rebootIdIt == _rebootIds.end()) {
+    std::string const error = [&]() {
+      std::stringstream strstream;
+      strstream << "When trying to register callback '" << callbackDescription << "': "
+                << "The server " << peerState.serverId() << " is not known. "
+                << "If this server joined the cluster in the last seconds, "
+                   "this can happen.";
+      return strstream.str();
+    }();
+    LOG_TOPIC("76abc", INFO, Logger::CLUSTER) << error;
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_CLUSTER_SERVER_UNKNOWN, error);
+  }
+
+  auto const currentRebootId = rebootIdIt->second;
+
+  if (peerState.rebootId() < currentRebootId) {
+    // If this ID is already older, schedule the callback immediately.
+    queueCallback(DescriptedCallback{std::move(callback), std::move(callbackDescription)});
+    return CallbackGuard{nullptr};
+  }
+
+  // For the given server, get the existing rebootId => [callbacks] map,
+  // or create a new one
+  auto& rebootIdMap = _callbacks[peerState.serverId()];
+  // For the given rebootId, get the existing callbacks map,
+  // or create a new one
+  auto& callbackMapPtr = rebootIdMap[peerState.rebootId()];
+
+  if (callbackMapPtr == nullptr) {
+    // We must never leave a nullptr in here!
+    // Try to create a new map, or remove the entry.
+    try {
+      callbackMapPtr =
+          std::make_shared<std::remove_reference<decltype(callbackMapPtr)>::type::element_type>();
+    } catch (...) {
+      rebootIdMap.erase(peerState.rebootId());
+      throw;
+    }
+  }
+
+  TRI_ASSERT(callbackMapPtr != nullptr);
+
+  auto& callbackMap = *callbackMapPtr;
+
+  auto const callbackId = getNextCallbackId();
+
+  // The guard constructor might, theoretically, throw, and so can constructing
+  // the std::function. So we need to construct it before emplacing the callback.
+  auto callbackGuard = CallbackGuard([this, peerState, callbackId]() {
+    unregisterCallback(peerState, callbackId);
+  });
+
+  auto emplaceRv =
+      callbackMap.emplace(callbackId, DescriptedCallback{std::move(callback),
+                                                         std::move(callbackDescription)});
+  auto const iterator = emplaceRv.first;
+  bool const inserted = emplaceRv.second;
+  TRI_ASSERT(inserted);
+  TRI_ASSERT(callbackId == iterator->first);
+
+  // TODO I'm wondering why this compiles (with clang, at least), as the copy
+  //      constructor is deleted. I don't think it should...
```

Comment:
Note to self: remove the comment. As copy elision is allowed here, the move constructor existing suffices.